packages/library-base/src/commonMain/kotlin/io/realm/kotlin/types/geo/GeoPolygon.kt

```diff
@@ -0,0 +1,130 @@
+ *   to indicate a closed ring (meaning you need at least 4 points to define the
+ *   polygon).
+ * - Rings may not cross, i.e. the boundary of a ring may not intersect both the interior and
+ *   exterior of any other ring.
+ * - Rings may not share edges, i.e. if a ring contains an edge AB, then no other ring may contain
+ *   AB or BA.
+ * - Rings may share vertices, however no vertex may appear twice in a single ring.
+ * - No ring may be empty.
+ *
+ * Holes may be nested inside each other, in which case a location will be considered "inside" the
+ * polygon if it is included in an odd number of rings, counting the outer ring as 1. For example,
+ * a polygon representing a square with side 10 centered at (0,0) with holes representing squares
+ * with sides 5 and 2, centered at (0,0) will  include the location (1, 1) because it is contained
+ * in 3 rings, but not (3, 3), because it is contained in 2.
+ *
+ * Using it can look something like this:
+ *
+ * ```
+ * val searchArea = GeoPolygon.create(
+ *     outerRing = listOf(
+ *         GeoPoint.create(0.0, 0.0),
+ *         GeoPoint.create(0.0, 5.0),
+ *         GeoPoint.create(5.0, 5.0),
+ *         GeoPoint.create(5.0, 0.0),
+ *         GeoPoint.create(0.0, 0.0),
+ *     ),
+ *     holes = arrayOf(listOf(
+ *         GeoPoint.create(1.0, 1.0),
+ *         GeoPoint.create(1.0, 4.0),
+ *         GeoPoint.create(4.0, 4.0),
+ *         GeoPoint.create(4.0, 1.0)
+ *     ))
+ * )
+ * val restaurants = realm.query<Restaurant>("location GEOWITHIN $searchArea").find()
+ * * ```
+ */
+@ExperimentalGeoSpatialApi
+public interface GeoPolygon {
+    /**
+     * A list of [GeoPoint]s representing the outer ring of the polygon.
+     */
+    public val outerRing: List<GeoPoint>
+
+    /**
+     * A list of inner rings, each ring is inside the next.
+     */
+    public val holes: List<List<GeoPoint>>
+
+    /**
+     * Returns the textual representation of the [GeoPolygon], this is also formatting it in a way
+     * that makes it usable in queries, e.g.:
+     *
+     * ```
+     * val outerRing = listOf(
+     *     GeoPoint.create(0.0, 0.0),
+     *     GeoPoint.create(10.0, 0.0),
+     *     GeoPoint.create(10.0, 10.0),
+     *     GeoPoint.create(0.0, 10.0),
+     *     GeoPoint.create(0.0, 0.0)
+     * )
+     * val searchArea = GeoPolygon.create(outerRing)
+     * val results = realm.query<Restaurant>("location GEOWITHIN searchArea").find()
+     */
+    override fun toString(): String
+
+    public companion object {
+
+        /**
+         * Create a [GeoPolygon] instance.
+         *
+         * @param outerRing A list of [GeoPoint]s representing the outer ring of the polygon. The
+         * outer ring must contain at least 3 unique points. The first and the last point may
+         * be identical, but no other duplicates are allowed. Each subsequent pair of points represents
+         * an edge in the polygon with the first and the last point being implicitly connected.
```

Comment:
A bit confused whether we need four with overlapping first/last (as the class doc states) or if we can define it with only three and consider the connection between first/last

Reply:
I think the confusion is that Core changed the behavior. But currently they do require a closed polygon where first and last point is the same. I'll modify the sentence here.